src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -53,53 +70,186 @@
+                }
+                if (!create) {
+                    return null;
+                }
+                // Allocate:
+                String mnemonic = mnemonicOf(fullName);
+                for (int i = 1; ; i++) {
+                    path = StringUtils.right(i > 1 ? mnemonic + "_" + i : mnemonic, MAX_LENGTH);
+                    if (!index.containsKey(path)) {
+                        dir = workspace.child(path);
+                        if (!dir.isDirectory()) {
+                            index.put(fullName, path);
+                            save(index, workspace);
+                            return dir;
+                        }
+                    }
+                }
             }
-            return new FilePath(new File(expandVariablesForDirectory(workspaceDir, minimized, item.getRootDir().getPath())));
-        } else if (node instanceof Slave) {
-            FilePath root = ((Slave) node).getWorkspaceRoot();
-            return root != null ? root.child(minimized) : null;
-        } else { // ?
+        } catch (IOException | InterruptedException x) {
+            LOGGER.log(Level.WARNING, "could not manage workspaces on " + node, x);
             return null;
         }
     }
+    
+    private static Map<String, String> load(FilePath workspace) throws IOException, InterruptedException {
+        Map<String, String> map = new TreeMap<>();
+        FilePath index = workspace.child(INDEX_FILE_NAME);
+        if (index.exists()) {
+            try (InputStream is = index.read(); Reader r = new InputStreamReader(is, StandardCharsets.UTF_8); BufferedReader br = new BufferedReader(r)) {
+                while (true) {
+                    String key = br.readLine();
+                    if (key == null) {
+                        break;
+                    }
+                    String value = br.readLine();
+                    if (value == null) {
+                        throw new IOException("malformed " + index);
+                    }
+                    map.put(key, value);
+                }
+            }
+        }
+        return map;
+    }
 
-    /** copied from {@link Jenkins} */
-    static String expandVariablesForDirectory(String base, String itemFullName, String itemRootDir) {
-        return Util.replaceMacro(base, ImmutableMap.of(
-                "JENKINS_HOME", Jenkins.getActiveInstance().getRootDir().getPath(),
-                "ITEM_ROOTDIR", itemRootDir,
-                "ITEM_FULLNAME", itemFullName,   // legacy, deprecated
-                "ITEM_FULL_NAME", itemFullName.replace(':','$'))); // safe, see JENKINS-12251
+    private static void save(Map<String, String> index, FilePath workspace) throws IOException, InterruptedException {
+        // FilePath.renameTo does not support REPLACE_EXISTING, and AtomicFileWriter does not work remotely.
+        // Anyway we are synchronizing access to this file so the only potential problem is half-written content.
```

Comment:
Maybe you could add something like the expected number of pairs to the first line of the file to help diagnose a corrupted index in `#load`, although with os/disk caches I'm not sure how likely it would be for that data to be written correctly if there is a problem with the rest of the file.